File: flutter/ios/Runner/Channels/CallChannelHandler.swift
import Foundation
import Flutter
import CallKit

/**
 * Platform channel handler for call management on iOS
 * Controls CallKit integration and bridges with Flutter
 */
class CallChannelHandler: NSObject {
    private let methodChannel: FlutterMethodChannel
    private let callKitManager = CallKitManager.shared

    init(messenger: FlutterBinaryMessenger) {
        methodChannel = FlutterMethodChannel(
            name: "com.callsafe.call",
            binaryMessenger: messenger
        )

        super.init()

        methodChannel.setMethodCallHandler(handle)
        setupCallbacks()
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "showIncomingCall":
            guard let args = call.arguments as? [String: Any],
                  let callAttemptId = args["callAttemptId"] as? String,
                  let callerName = args["callerName"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
                return
            }
            let isVideo = args["isVideo"] as? Bool ?? false
            showIncomingCall(callAttemptId: callAttemptId, callerName: callerName, isVideo: isVideo, result: result)

        case "startOutgoingCall":
            guard let args = call.arguments as? [String: Any],
                  let callAttemptId = args["callAttemptId"] as? String,
                  let recipientName = args["recipientName"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
                return
            }
            let isVideo = args["isVideo"] as? Bool ?? false
            startOutgoingCall(callAttemptId: callAttemptId, recipientName: recipientName, isVideo: isVideo, result: result)

        case "endCall":
            guard let args = call.arguments as? [String: Any],
                  let callAttemptId = args["callAttemptId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing callAttemptId", details: nil))
                return
            }
            endCall(callAttemptId: callAttemptId, result: result)

        case "setCallActive":
            guard let args = call.arguments as? [String: Any],
                  let callAttemptId = args["callAttemptId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing callAttemptId", details: nil))
                return
            }
            setCallActive(callAttemptId: callAttemptId, result: result)

        case "setMuted":
            guard let args = call.arguments as? [String: Any],
                  let callAttemptId = args["callAttemptId"] as? String,
                  let muted = args["muted"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
                return
            }
            setMuted(callAttemptId: callAttemptId, muted: muted, result: result)

        case "startForegroundService":
            // iOS doesn't need explicit foreground service, CallKit handles it
            result(nil)

        case "stopForegroundService":
            // iOS doesn't need explicit foreground service
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func setupCallbacks() {
        callKitManager.onAnswerCallback = { [weak self] callAttemptId in
            print("[CallChannel] Call answered: \(callAttemptId)")
            self?.methodChannel.invokeMethod("onCallAnswered", arguments: ["callAttemptId": callAttemptId])
        }

        callKitManager.onRejectCallback = { [weak self] callAttemptId in
            print("[CallChannel] Call rejected: \(callAttemptId)")
            self?.methodChannel.invokeMethod("onCallRejected", arguments: ["callAttemptId": callAttemptId])
        }

        callKitManager.onDisconnectCallback = { [weak self] callAttemptId in
            print("[CallChannel] Call disconnected: \(callAttemptId)")
            self?.methodChannel.invokeMethod("onCallDisconnected", arguments: ["callAttemptId": callAttemptId])
        }

        callKitManager.onMuteChangedCallback = { [weak self] callAttemptId, muted in
            print("[CallChannel] Mute changed: \(callAttemptId), muted=\(muted)")
            self?.methodChannel.invokeMethod("onMuteChanged", arguments: [
                "callAttemptId": callAttemptId,
                "muted": muted
            ])
        }
    }

    private func showIncomingCall(
        callAttemptId: String,
        callerName: String,
        isVideo: Bool,
        result: @escaping FlutterResult
    ) {
        callKitManager.reportIncomingCall(
            callAttemptId: callAttemptId,
            callerName: callerName,
            hasVideo: isVideo
        ) { error in
            if let error = error {
                result(FlutterError(
                    code: "SHOW_CALL_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            } else {
                result(nil)
            }
        }
    }

    private func startOutgoingCall(
        callAttemptId: String,
        recipientName: String,
        isVideo: Bool,
        result: @escaping FlutterResult
    ) {
        callKitManager.startOutgoingCall(
            callAttemptId: callAttemptId,
            recipientName: recipientName,
            hasVideo: isVideo
        ) { error in
            if let error = error {
                result(FlutterError(
                    code: "START_CALL_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            } else {
                result(nil)
            }
        }
    }

    private func endCall(callAttemptId: String, result: @escaping FlutterResult) {
        callKitManager.endCall(callAttemptId: callAttemptId) { error in
            if let error = error {
                result(FlutterError(
                    code: "END_CALL_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            } else {
                result(nil)
            }
        }
    }

    private func setCallActive(callAttemptId: String, result: @escaping FlutterResult) {
        callKitManager.setCallConnected(callAttemptId: callAttemptId)
        result(nil)
    }

    private func setMuted(callAttemptId: String, muted: Bool, result: @escaping FlutterResult) {
        callKitManager.setMuted(callAttemptId: callAttemptId, muted: muted) { error in
            if let error = error {
                result(FlutterError(
                    code: "MUTE_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            } else {
                result(nil)
            }
        }
    }

    func dispose() {
        methodChannel.setMethodCallHandler(nil)
        callKitManager.onAnswerCallback = nil
        callKitManager.onRejectCallback = nil
        callKitManager.onDisconnectCallback = nil
        callKitManager.onMuteChangedCallback = nil
    }
}

File: flutter/android/app/src/main/kotlin/com/callsafe/mobile/channels/CallChannelHandler.kt
package com.callsafe.mobile.channels

import android.content.ComponentName
import android.content.Context
import android.os.Build
import android.telecom.TelecomManager
import android.util.Log
import com.callsafe.mobile.call.CallConnectionService
import com.callsafe.mobile.call.CallForegroundService
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Platform channel handler for call management
 * Controls ConnectionService, foreground service, and call state
 */
class CallChannelHandler(
    messenger: BinaryMessenger,
    private val context: Context
) : MethodChannel.MethodCallHandler {

    private val channel = MethodChannel(messenger, CHANNEL_NAME)
    private val telecomManager = context.getSystemService(Context.TELECOM_SERVICE) as? TelecomManager

    companion object {
        private const val TAG = "CallChannelHandler"
        private const val CHANNEL_NAME = "com.callsafe.call"
    }

    init {
        channel.setMethodCallHandler(this)
        setupCallbacks()
        registerPhoneAccount()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            result.error("UNSUPPORTED", "ConnectionService requires Android 8.0+", null)
            return
        }

        when (call.method) {
            "showIncomingCall" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                val callerName = call.argument<String>("callerName")
                val isVideo = call.argument<Boolean>("isVideo") ?: false

                if (callAttemptId != null && callerName != null) {
                    showIncomingCall(callAttemptId, callerName, isVideo, result)
                } else {
                    result.error("INVALID_ARGS", "Missing required arguments", null)
                }
            }

            "startOutgoingCall" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                val recipientName = call.argument<String>("recipientName")
                val isVideo = call.argument<Boolean>("isVideo") ?: false

                if (callAttemptId != null && recipientName != null) {
                    startOutgoingCall(callAttemptId, recipientName, isVideo, result)
                } else {
                    result.error("INVALID_ARGS", "Missing required arguments", null)
                }
            }

            "endCall" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                if (callAttemptId != null) {
                    endCall(callAttemptId, result)
                } else {
                    result.error("INVALID_ARGS", "Missing callAttemptId", null)
                }
            }

            "setCallActive" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                if (callAttemptId != null) {
                    setCallActive(callAttemptId, result)
                } else {
                    result.error("INVALID_ARGS", "Missing callAttemptId", null)
                }
            }

            "startForegroundService" -> {
                val callAttemptId = call.argument<String>("callAttemptId")
                val callerName = call.argument<String>("callerName")
                if (callAttemptId != null && callerName != null) {
                    CallForegroundService.startService(context, callAttemptId, callerName)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Missing required arguments", null)
                }
            }

            "stopForegroundService" -> {
                CallForegroundService.stopService(context)
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    private fun registerPhoneAccount() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && telecomManager != null) {
            val componentName = ComponentName(context, CallConnectionService::class.java)
            CallConnectionService.registerPhoneAccount(telecomManager, componentName)
        }
    }

    private fun setupCallbacks() {
        CallConnectionService.onAnswerCallback = { callAttemptId ->
            Log.d(TAG, "Call answered: $callAttemptId")
            channel.invokeMethod("onCallAnswered", mapOf("callAttemptId" to callAttemptId))
        }

        CallConnectionService.onRejectCallback = { callAttemptId ->
            Log.d(TAG, "Call rejected: $callAttemptId")
            channel.invokeMethod("onCallRejected", mapOf("callAttemptId" to callAttemptId))
        }

        CallConnectionService.onDisconnectCallback = { callAttemptId ->
            Log.d(TAG, "Call disconnected: $callAttemptId")
            channel.invokeMethod("onCallDisconnected", mapOf("callAttemptId" to callAttemptId))
            CallForegroundService.stopService(context)
        }

        CallConnectionService.onMuteChangedCallback = { callAttemptId, muted ->
            Log.d(TAG, "Mute changed: $callAttemptId, muted=$muted")
            channel.invokeMethod(
                "onMuteChanged",
                mapOf("callAttemptId" to callAttemptId, "muted" to muted)
            )
        }

        CallConnectionService.onSpeakerChangedCallback = { callAttemptId, speaker ->
            Log.d(TAG, "Speaker changed: $callAttemptId, speaker=$speaker")
            channel.invokeMethod(
                "onSpeakerChanged",
                mapOf("callAttemptId" to callAttemptId, "speaker" to speaker)
            )
        }
    }

    private fun showIncomingCall(
        callAttemptId: String,
        callerName: String,
        isVideo: Boolean,
        result: MethodChannel.Result
    ) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && telecomManager != null) {
            try {
                val componentName = ComponentName(context, CallConnectionService::class.java)
                CallConnectionService.showIncomingCall(
                    telecomManager,
                    componentName,
                    callAttemptId,
                    callerName,
                    isVideo
                )
                result.success(null)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to show incoming call", e)
                result.error("SHOW_CALL_FAILED", e.message, null)
            }
        } else {
            result.error("UNSUPPORTED", "TelecomManager not available", null)
        }
    }

    private fun startOutgoingCall(
        callAttemptId: String,
        recipientName: String,
        isVideo: Boolean,
        result: MethodChannel.Result
    ) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O && telecomManager != null) {
            try {
                val componentName = ComponentName(context, CallConnectionService::class.java)
                CallConnectionService.startOutgoingCall(
                    telecomManager,
                    componentName,
                    callAttemptId,
                    recipientName,
                    isVideo
                )
                result.success(null)
            } catch (e: Exception) {
                Log.e(TAG, "Failed to start outgoing call", e)
                result.error("START_CALL_FAILED", e.message, null)
            }
        } else {
            result.error("UNSUPPORTED", "TelecomManager not available", null)
        }
    }

    private fun endCall(callAttemptId: String, result: MethodChannel.Result) {
        CallConnectionService.endCall(callAttemptId)
        CallForegroundService.stopService(context)
        result.success(null)
    }

    private fun setCallActive(callAttemptId: String, result: MethodChannel.Result) {
        CallConnectionService.setCallActive(callAttemptId)
        result.success(null)
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
        CallConnectionService.onAnswerCallback = null
        CallConnectionService.onRejectCallback = null
        CallConnectionService.onDisconnectCallback = null
        CallConnectionService.onMuteChangedCallback = null
        CallConnectionService.onSpeakerChangedCallback = null
    }
}
